rclpy/rclpy/parameter.py

```diff
@@ -284,36 +284,60 @@ def parameter_dict_from_yaml_file(
     """
     with open(parameter_file, 'r') as f:
         param_file = yaml.safe_load(f)
-        param_keys = []
         param_dict = {}
 
-        if use_wildcard and '/**' in param_file:
-            param_keys.append('/**')
+        # check and add if wildcard is available in 1st keys
+        # wildcard key must go to the front of param_keys so that
+        # node-namespaced parameters will override the wildcard parameters
+        if use_wildcard and '/**' in param_file.keys():
+            value = param_file['/**']
+            if type(value) != dict and 'ros__parameters' not in value:
+                raise RuntimeError(
+                    'YAML file is not a valid ROS parameter file for wildcard(/**)')
+            param_dict.update(value['ros__parameters'])
 
+        # parse parameter yaml file based on target node namespace and name
         if target_nodes:
             for n in target_nodes:
-                if n not in param_file.keys():
-                    raise RuntimeError(f'Param file does not contain parameters for {n},'
-                                       f'only for nodes: {list(param_file.keys())} ')
-                param_keys.append(n)
-        else:
-            # wildcard key must go to the front of param_keys so that
-            # node-namespaced parameters will override the wildcard parameters
-            keys = set(param_file.keys())
-            keys.discard('/**')
-            param_keys.extend(keys)
-
-        if len(param_keys) == 0:
-            raise RuntimeError('Param file does not contain selected parameters')
-
-        for n in param_keys:
-            value = param_file[n]
-            if type(value) != dict or 'ros__parameters' not in value:
-                raise RuntimeError(f'YAML file is not a valid ROS parameter file for node {n}')
-            param_dict.update(value['ros__parameters'])
+                abs_name = _get_absolute_node_name(n)
+                ns, node_basename = abs_name.rsplit('/', 1)
```

Comment:
I agree that we should maintain the same behavior as `rcl_param_yaml_parser`.

But I'd actually like to back-up a bit and look at `parameter_dict_from_yaml_file`.  In particular, it has 1 required argument (`parameter_file`), and 3 optional arguments (`use_wildcard`, `target_nodes`, and `namespace`).  I took a look through the core and through all of the code released into Rolling, and there are only a handful of non-test calls to this function:

* https://github.com/ros2/ros2cli/blob/419afea1282ae4daa43ea446a2c0bcc7394f856c/ros2param/ros2param/api/__init__.py#L42
* https://github.com/ros2/demos/blob/a29e65623e8ace15b8253ab33f19f3d8165c3cfe/demo_nodes_py/demo_nodes_py/parameters/async_param_client.py#L76
* https://github.com/ros2/rclpy/blob/e97c41b019c6deb1c490a7435164d1a38bcaac27/rclpy/rclpy/parameter_client.py#L312
* https://github.com/ros2/rclpy/blob/e97c41b019c6deb1c490a7435164d1a38bcaac27/rclpy/rclpy/parameter_client.py#L334

We never use the `target_nodes` or the `namespace` arguments at all.

So my question is: should we just remove these parameters and the functionality around them?  It would make this function much easier to understand and test.  But maybe I am missing something, I'd like to hear your thoughts.

Reply:
@clalancette thanks!

- https://github.com/ros2/ros2cli/blob/419afea1282ae4daa43ea446a2c0bcc7394f856c/ros2param/ros2param/api/__init__.py#L42

`ros2 param load <node> <parameter_file>` case, i think we need to use `target_nodes` parameter to specify `<node>` with `namespace` from the yaml file. this is the original issue described https://github.com/ros2/ros2cli/issues/863, wrong parameters are loaded to the `<node>`.

- https://github.com/ros2/demos/blob/a29e65623e8ace15b8253ab33f19f3d8165c3cfe/demo_nodes_py/demo_nodes_py/parameters/async_param_client.py#L76

this can work at this moment, only because it depends on the bug which this PR trying to address.

we can see, https://github.com/ros2/demos/blob/a29e65623e8ace15b8253ab33f19f3d8165c3cfe/demo_nodes_py/demo_nodes_py/parameters/params.yaml#L1

parameter yaml file only has `/param_test_target` absolute node name, which should never be loaded to

https://github.com/ros2/demos/blob/a29e65623e8ace15b8253ab33f19f3d8165c3cfe/demo_nodes_py/demo_nodes_py/parameters/async_param_client.py#L28

`parameter_blackboard`, because yaml file does not have `parameter_blackboard` parameters at all.

- https://github.com/ros2/rclpy/blob/e97c41b019c6deb1c490a7435164d1a38bcaac27/rclpy/rclpy/parameter_client.py#L312
- https://github.com/ros2/rclpy/blob/e97c41b019c6deb1c490a7435164d1a38bcaac27/rclpy/rclpy/parameter_client.py#L334

`ParameterClient` created on `remote node`, so when loading we can validate the parameters from the yaml file what needs to be loaded against `remote node`.

> We never use the target_nodes or the namespace arguments at all.

`parameter_dict_from_yaml_file` can just convert yaml into dictionary, and then caller side can do filtering based on those `target_nodes`, `namespace` and `wildcard`, but i do not think that is really useful. filtering logic should be implemented in `parameter_dict_from_yaml_file`. so that we can get what needs to be loaded based on target node names, namespace and wildcard. (i guess that was original intention for this `parameter_dict_from_yaml_file` but not tested very good...)

Reply:
OK, I see.  Sorry, I was looking at the original code and not this PR when I wrote my previous comment.

Let me back up even further and ask a question.  Is the intention for `parameter_dict_from_yaml_file` to be the Python equivalent of https://github.com/ros2/rcl/blob/246fd1de5c073e60573abb014a787f012a241419/rcl_yaml_param_parser/include/rcl_yaml_param_parser/parser.h#L96-L98 ?

Reply:
I am not sure the original intention, sorry. i was just trying to fix https://github.com/ros2/ros2cli/issues/863, and also check if it has the same behavior with `rcl_yaml_param_parser`. not sure if this answers your question though...

Reply:
@clalancette friendly ping. can we resolve this comment?